### DISCUSSION TITLE
Instance has been deleted error after commit

### DISCUSSION THREAD

**Discussion by rajendrakhabiyagate6** [sqlalchemy/sqlalchemy#10233] [Usage Questions]
I have a Read-Write replication setup on MYSQL database and I am redirecting all Select queries on Read instance and all other on Write using RoutingSession class as given below :

```
# create engine
engines = {
    "read": create_engine(readConnectionString, convert_unicode=True, poolclass=NullPool,isolation_level="READ UNCOMMITTED", pool_recycle = 800),
    "write": create_engine(writeConnectionString, convert_unicode=True, poolclass=NullPool,isolation_level="READ UNCOMMITTED", pool_recycle = 800)
}

class RoutingSession(Session):     
    def get_bind(self, mapper=None, clause=None):  
        if isinstance(clause, (Select)):
            return engines["read"]
        else:
            return engines["write"]
                
Session = sessionmaker(class_=RoutingSession, autocommit=False, autoflush=True)
db_session = scoped_session(Session)
```

Everything is working fine, it is routing all Select queries on Read engine and all other on Write engine. But sometimes when we write anything using db_session.commit() and try to get any column from the instance after commit it gives and error like this 

```
UserObj = User(None, userName, userContact)
db_session.add(UserObj)
db_session.commit()
print(User.userId)  #just print Primary Key of last insert

Got this error
**Instance '<User at 0x7fdaf22fffa0>' has been deleted, or its row is otherwise not present**

```

While we check log using echo we found that a select is executed after commit and which routed to Read engine and give this error. Not sure how to handle such situation ? Any help would be appreciated.

**Comment by CaselIT** [comment]
Hi,

It seems that the write has not yet been replicated to the replica where the select happens.

**Reply by rajendrakhabiyagate6** [reply]
Not sure because there is no lag in replication when i checked

**Reply by CaselIT** [reply]
If you force the use of the writer engine, does the error disappears?

**Reply by rajendrakhabiyagate6** [reply]
yes, If I use only write engine for all operations never get this error

**Reply by CaselIT** [reply]
Then the issue is likely replication delay

**Reply by CaselIT** [reply]
Try putting a sleep before the select to confirm that's the case

**Reply by rajendrakhabiyagate6** [reply]
So what would be the best approach to handle such case using sqlalchemy, any suggestions ?

**Reply by CaselIT** [reply]
If I'm not mistaken you can pass arguments to the get bind method, so you could force the use of the writer engine after adding new objects to a session.

Or you can use the session.info dict

**Reply by rajendrakhabiyagate6** [reply]
Can you please help me with get_bind method so that after write i can use same engine for next select

**Reply by CaselIT** [reply]
The easiest solution is probably something like this

```py
class StickyRoutingSession(Session):     
    def get_bind(self, mapper=None, clause=None):  
        if not isinstance(clause, Select) or self info.get('use_write'):
            self.info['use_write'] = True
            return engines["write"]
        else:
            return engines["read"]

**Reply by rajendrakhabiyagate6** [reply]
It doesn't suits because if there are other operation after commit and before getting value all routed to Write engine

**Reply by CaselIT** [reply]
You will have to manually tell the session what engine to use then

**Comment by zzzeek** [comment]
in my experience with this use case, even though the `RoutingSession` is up on the site I dont actually use that approach in practice.   IMO the direction to reader/writer should be done on a per-transaction basis.  This means you have **two** `sessionmaker()` factories, and for a given top-level request, if the request is read only, you use the read-only session, if the request includes writes, you use the writer session.  this makes the application much easier to work with and also ensures that all the data you are consuming within a "writer" transaction is correctly aligned transaction-wise with things that have been written.

**Reply by zzzeek** [reply]
I've just updated the docs at d5e886b282d838b9920175 to push people towards the [multiple session example](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-isolation-for-a-sessionmaker-engine-wide) instead

**Reply by rajendrakhabiyagate6** [reply]
@zzzeek Thanks, but I don't understand it can you please help me with a proper example similar to my case. Like do we have to create two database sessions and query separately like below :

```
#create engine
engines = {
    "read": create_engine(readConnectionString, convert_unicode=True, poolclass=NullPool,isolation_level="READ UNCOMMITTED", pool_recycle = 800),
    "write": create_engine(writeConnectionString, convert_unicode=True, poolclass=NullPool,isolation_level="READ UNCOMMITTED", pool_recycle = 800)
}

#Read Session for Select queries
Session_Read = sessionmaker(bind=engines["read"], autocommit=False, autoflush=True)
db_session_read = scoped_session(Session_Read)

 #Write Session for Insert, Update, Delete
Session_Write = sessionmaker(bind=engines["write"], autocommit=False, autoflush=True)
db_session_write = scoped_session(Session_Write)

```
then queries can be executed like this 

```
#Insert Sample
UserObj = User(None, userName, userContact)
db_session_write.add(UserObj)
db_session_write.commit()
print(User.userId)  #just print Primary Key of last insert

#Select Sample
user_list = "Select * From Users"
result = db_session_read.execute(text(user_list)).fetchall()

```

**Reply by zzzeek** [reply]
yes you have to mark operations at the top of your transactions for "reader" or "writer".  this typically applies best to web applications where specific web request methods can be marked as "reader" or "writer" based on something like HTTP method (e.g. a GET is a "reader", a POST is a "writer", for example).    In practice you might be writing things like logs etc. in GET methods, so a single request might use both "reader" and "writer" for those separate tasks

such as:

```py
def read_some_data():
    with reader_session() as sess:
        data1= get_some_data_one(sess)
        data2= get_some_data_two(sess)
        # .. .etc

def write_some_data(data):
    with writer_session() as session:
          run_method_one(session, data)
          run_method_two(session, data)
          # ... etc
```

the idea is that the database itself has a transactional model, and if you are doing SELECTs for rows that are related to some INSERT, UPDATE, DELETE you are about to do, you want all that in one transaction.  you wouldn't want to break that up among two different connections.   so the `RoutingSession` example is not actually as useful as it immediately seems.
